---
title: "Assignment9"
author: "Liam McCrea"
date: "2023-11-08"
output: html_document
---


```{r setup}

#BiocManager::install("biomformat")
library(metagenomeSeq)
library(biomformat)

```

# Introduction

Metagenomeseq is a package that accounts for microbial marker gene bias in previous analysis methods.
It includes a novel normalization technique and addresses undersampling.
The analysis will determine features that are differentially abundant between groups of samples.

Steps:
  - prepare object
  - Normalize
  - Statistical testing
  - visualization


# Data Preparation

  - cluster reads according to sequence similarity and define counts using algorithms
  - the data is formatted as a MRexperiment object.

## Biom Format

  - start with biom file format data to transition to mrexperiment 


```{r Biom format}

biom_file <- system.file("extdata", "min_sparse_otu_table.biom",
package = "biomformat")
b <- read_biom(biom_file)
biom2MRexperiment(b)

# example writing biom file
data(mouseData)
# options include to normalize or not
b <- MRexperiment2biom(mouseData)
write_biom(b, biom_file = "otu_table.biom")


```

## Loading count data

```{r loading count data}

dataDirectory <- system.file("extdata", package = "metagenomeSeq")
lung = loadMeta(file.path(dataDirectory, "CHK_NAME.otus.count.csv"))
dim(lung$counts)

```

## loading taxonomy

add annotations for taxonomy.

```{r }

taxa = read.delim(file.path(dataDirectory, "CHK_otus.taxonomy.csv"),
stringsAsFactors = FALSE)

```

## loading metadata

Add list of other associated data.

```{r}

clin = loadPhenoData(file.path(dataDirectory, "CHK_clinical.csv"),
tran = TRUE)
ord = match(colnames(lung$counts), rownames(clin))
clin = clin[ord, ]
head(clin[1:2, ])

```

## Creating a MRexperiment object

ingest count matrix and associated annotations dataframes.

```{r}

phenotypeData = AnnotatedDataFrame(clin)
phenotypeData

OTUdata = AnnotatedDataFrame(taxa)
OTUdata

obj = newMRexperiment(lung$counts,phenoData=phenotypeData,featureData=OTUdata)
# Links to a paper providing further details can be included optionally.
# experimentData(obj) = annotate::pmid2MIAME("21680950")
obj

```

## Example datasets

human lung microbiome and humanized gnotobiotic mouse gut are example datasets in the package.

```{r}

data(lungData)
lungData

data(mouseData)
mouseData
```

## Useful commands

The following are examples of how you can check the input data.

```{r}
phenoData(obj)
head(pData(obj), 3)
featureData(obj)
head(fData(obj)[, -c(2, 10)], 3)
head(MRcounts(obj[, 1:2]))

featuresToKeep = which(rowSums(obj) >= 100)
samplesToKeep = which(pData(obj)$SmokingStatus == "Smoker")
obj_smokers = obj[featuresToKeep, samplesToKeep]
obj_smokers

head(pData(obj_smokers), 3)

head(normFactors(obj))

normFactors(obj) <- rnorm(ncol(obj))
head(normFactors(obj))

head(libSize(obj))

libSize(obj) <- rnorm(ncol(obj))
head(libSize(obj))

data(mouseData)
filterData(mouseData, present = 10, depth = 1000)



```


# Normalization

Required due to varying levels of coverage.
need to carefully set the parameters and record used parameter.

## Calculating normalizing factors

calculate proper percentile, scaling factors.

```{r}

data(lungData)
p = cumNormStatFast(lungData)
lungData = cumNorm(lungData, p = p)
head(lungData)

```

## calculating normalization factors using Wrench

An alternative approach preferred for cumulative normalization.

```{r}

condition = mouseData$diet
mouseData = wrenchNorm(mouseData, condition = condition)

```

## Exporting data

Export now that its normalized.

```{r}

mat = MRcounts(lungData, norm = TRUE, log = TRUE)[1:5, 1:5]
exportMat(mat, file = file.path(dataDirectory, "tmp.tsv"))

exportStats(lungData[, 1:5], file = file.path(dataDirectory,
"tmp.tsv"))
## Default value being used.
head(read.csv(file = file.path(dataDirectory, "tmp.tsv"), sep = "\t"))


```



# Statistical testing

Now that its normalized we can compare groups - differentially abundant features.
This includes determining operational taxonomic unit (OTUs), species, Genes, etc.
There are multiple model options but you must define the appropriate model for testing.

## fitFeatureModel for abundance testing

```{r}

data(lungData)
lungData = lungData[, -which(is.na(pData(lungData)$SmokingStatus))]
lungData = filterData(lungData, present = 30, depth = 1)
lungData <- cumNorm(lungData, p = 0.5)
pd <- pData(lungData)
mod <- model.matrix(~1 + SmokingStatus, data = pd)
lungres1 = fitFeatureModel(lungData, mod)
head(MRcoefs(lungres1))

```


## fitZig for differential abundance 


```{r}

data(lungData)
controls = grep("Extraction.Control", pData(lungData)$SampleType)
lungTrim = lungData[, -controls]
rareFeatures = which(rowSums(MRcounts(lungTrim) > 0) < 10)
lungTrim = lungTrim[-rareFeatures, ]
lungp = cumNormStat(lungTrim, pFlag = TRUE, main = "Trimmed lung data")
## Default value being used.

lungTrim = cumNorm(lungTrim, p = lungp)

```


```{r}

smokingStatus = pData(lungTrim)$SmokingStatus
bodySite = pData(lungTrim)$SampleType
normFactor = normFactors(lungTrim)
normFactor = log2(normFactor/median(normFactor) + 1)
mod = model.matrix(~smokingStatus + bodySite + normFactor)
settings = zigControl(maxit = 10, verbose = TRUE)
fit = fitZig(obj = lungTrim, mod = mod, useCSSoffset = FALSE,
control = settings)

```


## Multiple groups

```{r}

# maxit=1 is for demonstration purposes
settings = zigControl(maxit = 1, verbose = FALSE)
mod = model.matrix(~bodySite)
colnames(mod) = levels(bodySite)
# fitting the ZIG model
res = fitZig(obj = lungTrim, mod = mod, control = settings)
# The output of fitZig contains a list of various useful
# items. hint: names(res). Probably the most useful is the
# limma 'MLArrayLM' object called fit.
zigFit = slot(res, "fit")
finalMod = slot(res, "fit")$design
contrast.matrix = makeContrasts(BAL.A - BAL.B, OW - PSB, levels = finalMod)
fit2 = contrasts.fit(zigFit, contrast.matrix)
fit2 = eBayes(fit2)
topTable(fit2)


```

## Exporting

```{r}

taxa = sapply(strsplit(as.character(fData(lungTrim)$taxa), split = ";"),
function(i) {
i[length(i)]
})
head(MRcoefs(fit, taxa = taxa, coef = 2))


```

## Log normal permutation test

```{r}

coeffOfInterest = 2
res = fitLogNormal(obj = lungTrim, mod = mod, useCSSoffset = FALSE,
B = 10, coef = coeffOfInterest)
# extract p.values and adjust for multiple testing res$p
# are the p-values calculated through permutation
adjustedPvalues = p.adjust(res$p, method = "fdr")
# extract the absolute fold-change estimates
foldChange = abs(res$fit$coef[, coeffOfInterest])
# determine features still significant and order by the
sigList = which(adjustedPvalues <= 0.05)
sigList = sigList[order(foldChange[sigList])]
# view the top taxa associated with the coefficient of
# interest.
head(taxa[sigList])


```

## presence absence testing

```{r}
classes = pData(mouseData)$diet
res = fitPA(mouseData[1:5, ], cl = classes)
# Warning - the p-value is calculating 1 despite a high
# odd's ratio.
head(res)

```

## Discovery odds ratio testing

```{r}

classes = pData(mouseData)$diet
res = fitDO(mouseData[1:100, ], cl = classes, norm = FALSE, log = FALSE)
head(res)

```

## Feature correlations

now that we have the abundance features we can pairwise compare them

```{r}

cors = correlationTest(mouseData[55:60, ], norm = FALSE, log = FALSE)
head(cors)

```

## Unique OTUs

find unique features

```{r}

cl = pData(mouseData)[["diet"]]
uniqueFeatures(mouseData, cl, nsamples = 10, nreads = 100)

```

# Aggregating counts

can combine counts to a greater level, whatever it is - genus, taxon, etc

```{r}

obj = aggTax(mouseData, lvl = "phylum", out = "matrix")
head(obj[1:5, 1:5])


```

```{r}

obj = aggSamp(mouseData, fct = "mouseID", out = "matrix")
head(obj[1:5, 1:5])

```


# Visualization

contains built in plotting functions

## structural overview

```{r}

trials = pData(mouseData)$diet
heatmapColColors = brewer.pal(12, "Set3")[as.integer(factor(trials))]
heatmapCols = colorRampPalette(brewer.pal(9, "RdBu"))(50)
plotMRheatmap(obj = mouseData, n = 200, cexRow = 0.4, cexCol = 0.4,
trace = "none", col = heatmapCols, ColSideColors = heatmapColColors)
```


```{r}

# plotCorr
plotCorr(obj = mouseData, n = 200, cexRow = 0.25, cexCol = 0.25,
trace = "none", dendrogram = "none", col = heatmapCols)

```

```{r}

cl = factor(pData(mouseData)$diet)
# plotOrd - can load vegan and set distfun = vegdist and
# use dist.method='bray'
plotOrd(mouseData, tran = TRUE, usePCA = FALSE, useDist = TRUE,
bg = cl, pch = 21)
# plotRare
res = plotRare(mouseData, cl = cl, pch = 21, bg = cl)
# Linear fits for plotRare / legend
tmp = lapply(levels(cl), function(lv) lm(res[, "ident"] ~ res[,
"libSize"] - 1, subset = cl == lv))
for (i in 1:length(levels(cl))) {
abline(tmp[[i]], col = i)
}
legend("topleft", c("Diet 1", "Diet 2"), text.col = c(1, 2),
box.col = NA)

```


## feature specific

```{r}

head(MRtable(fit, coef = 2, taxa = 1:length(fData(lungTrim)$taxa)))

```


```{r}

patients = sapply(strsplit(rownames(pData(lungTrim)), split = "_"),
function(i) {
i[3]
})
pData(lungTrim)$patients = patients
classIndex = list(smoker = which(pData(lungTrim)$SmokingStatus ==
"Smoker"))
classIndex$nonsmoker = which(pData(lungTrim)$SmokingStatus ==
"NonSmoker")
otu = 779
# plotOTU
plotOTU(lungTrim, otu = otu, classIndex, main = "Neisseria meningitidis")
# Now multiple OTUs annotated similarly
x = fData(lungTrim)$taxa[otu]
otulist = grep(x, fData(lungTrim)$taxa)
# plotGenus
plotGenus(lungTrim, otulist, classIndex, labs = FALSE, main = "Neisseria meningitidis")
lablist <- c("S", "NS")
axis(1, at = seq(1, 6, by = 1), labels = rep(lablist, times = 3))

```


```{r}

classIndex = list(Western = which(pData(mouseData)$diet == "Western"))
classIndex$BK = which(pData(mouseData)$diet == "BK")
otuIndex = 8770
# par(mfrow=c(1,2))
dates = pData(mouseData)$date
plotFeature(mouseData, norm = FALSE, log = FALSE, otuIndex, classIndex,
col = dates, sortby = dates, ylab = "Raw reads")

```


# Summary

Designed for marker-gene survey datasets - sparse dataset - and employs novel normalization with zero-inflated Gaussian mixture model.

## citation
```{r}

citation("metagenomeSeq")

```

## session info

records packages, parameters, etc used.

```{r}

sessionInfo()

```




